File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
import java.util.*;

class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        
        Arrays.sort(arrayA);
        Arrays.sort(arrayB);
        
        List<Integer> successA = isSucceedToDivide(arrayA);
        int ans1 = isFailedToDivide(arrayB, successA);
        
        List<Integer> successB = isSucceedToDivide(arrayB);
        int ans2 = isFailedToDivide(arrayA, successB);
        
        return Math.max(ans1, ans2);
    }
    
    private List<Integer> isSucceedToDivide(int[] array) {
        
        List<Integer> successList = new ArrayList<>();
        
        for (int i = array[0]; i >= 2; i--) {
            
            boolean success = true;
            
            for (int num : array) {
                if (num % i != 0) {
                    success = false;
                    break;
                }
            }
            
            if (success) {
                successList.add(i);
            }
        }
        
        return successList;
    }
    
    private int isFailedToDivide(int[] array, List<Integer> successList) {
        
        for (int i : successList) {
            
            boolean failure = true;
            
            for (int num : array) {
                
                if (num % i == 0) {
                    failure = false;
                    break;
                }
            }
            
            if (failure) {
                return i;
            }
        }
        return 0;
    }
}

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> isSucceedToDivide(vector<int> array) {
    
    vector<int> successList;
    
    for (int i = array[0]; i >= 2; i--) {
        
        bool success = true;
        
        for (int num : array) {
            
            if (num % i != 0) {
                success = false;
                break;
            }
        }
        
        if (success) {
            successList.push_back(i);
        }
    }
    
    return successList;
}

int isFailedToDivide(vector<int> array, vector<int> successList) {
    
    for (int i : successList) {
        bool failure = true;
        
        for (int num : array) {
            if (num % i == 0) {
                failure = false;
                break;
            }
        }
        
        if (failure) {
            return i;
        }
    }
    return 0;
}

int solution(vector<int> arrayA, vector<int> arrayB) {

    sort(arrayA.begin(), arrayA.end());
    sort(arrayB.begin(), arrayB.end());
    
    vector<int> successList = isSucceedToDivide(arrayA);
    int ans1 = isFailedToDivide(arrayB, successList);
        
    successList = isSucceedToDivide(arrayB);
    int ans2 = isFailedToDivide(arrayA, successList);
    
    return max(ans1, ans2);
}